### ISSUE TITLE
[Question] How can I flatten a ResultAsync in error case? 

### ISSUE THREAD

**Issue by Hyzual** [supermacro/neverthrow#282]
Hi, first of all thank you very much for writing and sharing `neverthrow`, it's changed my way of thinking about errors 😊.

I have a situation where I don't know how to unwrap and flatten a `ResultAsync<T, <ResultAsync<string, string>>`. Ideally, I would prefer to have something like `ResultAsync<T, string>`. I understand that `.andThen()` can help you flatten nested `ResultAsync`s but I did not manage to do it in the error case (not ok case). 

Consider an example like using the `fetch` API. In case of error, you might want to read the response's JSON or do another query. Code could look like this.

```typescript
import { fromPromise, ResultAsync, errAsync } from "neverthrow";

function getAPI(id: string): ResultAsync<Report, ResultAsync<string, string>> {
  return fromPromise(() => remoteAPICall(id), (error) => {
    if (error instanceof FetchWrapperError) {
      return fromPromise(() => error.response.json()), () => "Could not decode Error Json")
     .map((json) => json.error.message);
    }
    return errAsync("Other error");
 })
```

In my example above, in case of a rejected promise from `remoteAPICall()`, I need to evaluate a second promise `response.json()`. My example would not pass type-checking as I'm typing from memory, but I wanted to convey the idea of why I want to flatten nested `ResultAsync` in error case. I can figure out how to flatten nested promises by chaining returns but I don't know how to do it for nested `ResultAsync`. I want to have only a single `ResultAsync<Report, string>` as return type. 

Is it possible? Is there a way? Some syntax in neverthrow that could let me do this? I need your help 🙇‍♂️

Apologies for the mistakes or bad indentation, I'm typing this on my phone. I'll try to provide a running example tomorrow 😅

**Comment by supermacro** [comment]
Hey @Hyzual,

Thank you! I'm happy that you're enjoying `neverthrow` 😃 

Hmm ok, I think I understand what you're trying to do. Let me rephrase to see if I understand:

- wrap a function that returns `Promise<T>` inside of `fromPromise`
- This means that now we have a `ResultAsync<T, SomeType>`, where `SomeType` depends on the return of the `fromPromise` callback
- The `fromPromise` callback itself returns a `ResultAsync<U, E>` ... which now means you have a `ResultAsync<T, ResultAsync<U, E>>
- In your case `U` and `E` are strings from the looks of it?
- So you want `U` and `E` (which, again, are both `string`) to be unwrapped so that you end up with `Result<T, string>`

Re-reading what you said and comparing with what I wrote - I think I'm understanding correctly ... but let me know if I got something wrong!

I think there may be a way to use [`ResultAsync.orElse`](https://github.com/supermacro/neverthrow#resultasyncorelse-method):

```
// I have annotated things based on your specified types ... although I haven't actually verified that this compiles

ResultAsync<Report, ResultAsync<string, string>>.orElse(
  (error: ResultAsync<string, string>) =>
    error.match(
       (left) => left,
       (right) => right,
    )
)
```


Although this seems a little bit ... unnecessary. A more practical approach would be to bind `.catch` to your fetch api call and do your `if (error instanceof FetchWrapperError) {` check in the `.catch` callback before wrapping all of that code inside of `fromPromise`. 

Let me know if this helped unblock you 🙂

**Comment by Hyzual** [comment]
Hi @supermacro, thank you for the quick and complete response ! :bow: 
Indeed you understood correctly, that is what I wanted to do.
You're right, what I really needed was just chaining the Promises in the first place by binding `.catch` and only _then_ pass a promise to `fromPromise`.

Posting the code here in case anyone looks for a solution:

```typescript
import type { Report } from "./types";
import { ResultAsync } from "neverthrow";

function handleFetchWrapperError(error: FetchWrapperError): Promise<string> {
    // If the response is not even JSON
    if (typeof error.response.json !== "function") {
        // Return a rejected promise so that it will end up in the "Error" side of ResultAsync.
        return Promise.reject("Could not decode Error JSON");
    }
    return error.response.json().then(
        (error_json) => {
            if (
                Object.prototype.hasOwnProperty.call(error_json, "error") &&
                Object.prototype.hasOwnProperty.call(error_json.error, "message")
            ) {
                // Return a rejected promise so that it will end up in the "Error" side of ResultAsync
                // If we just return a string, since we are in a then(), the Promise 
                // would be resolved (and not rejected)
                return Promise.reject(error_json.error.message);
            }
            // Return a rejected promise so that it will end up in the "Error" side of ResultAsync
            return Promise.reject("Could not read error message from response");
        },
        () => {
            // This is the "catch" handler, so we can simply return a string, it will become a Promise<string>
            return "Could not decode Error JSON";
        }
    );
}

function getAPIAndHandleFetchWrapperError(uri: string): Promise<Response> {
    return get(uri).catch((error) => {
        // FetchWrapperError is a subclass of Error
        if (error instanceof FetchWrapperError) {
            // Using "return" here to chain the Promises and have only one Promise
            return handleFetchWrapperError(error);
        }
        return error;
    });
}

function getAPI(uri: string): ResultAsync<Report, string> {
    return ResultAsync.fromPromise(getAPIAndHandleFetchWrapperError(uri), (error: unknown) => {
        // Make sure the error is of type "string" to match the type of ResultAsync
        if (typeof error === "string") {
            return error;
        }
        return "Error in GET API";
    });
}

/* -------------------------------
calling code in another file
*/

async loadBackendReport(): Promise<void> {
    const result = await getReport(this.report_id);      
    if (result.isOk()) {
        //some stuff
    } else {
        // We don't need to await result.error because promises have been chained.
        const error_message = result.error; 
        // Display error_message
    }
}
```


Thank you very much, it did unblock me 🙂
